ambry-commons/src/main/java/com.github.ambry.commons/HelixNotifier.java

```diff
@@ -0,0 +1,209 @@
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final String topicPath;
+  private final HelixPropertyStore<ZNRecord> helixStore;
+  private final Map<TopicListener, HelixPropertyListener> topicListenerToHelixListenerMap = new ConcurrentHashMap<>();
+  private final Map<String, List<TopicListener>> topicToListenerMap = new ConcurrentHashMap<>();
+
+  /**
+   * Constructor.
+   * @param helixConfig The {@link HelixPropertyStoreConfig} that specifies necessary configs to start a HelixMessenger.
+   * @param helixPropertyStoreFactory The factory to start a {@link HelixPropertyStore}.
+   */
+  public HelixNotifier(HelixPropertyStoreConfig helixConfig,
+      HelixPropertyStoreFactory<ZNRecord> helixPropertyStoreFactory) {
+    this(helixConfig.rootPath, helixConfig.topicPath, helixConfig.zkClientConnectString,
+        helixConfig.zkClientSessionTimeoutMs, helixConfig.zkClientConnectionTimeoutMs, helixPropertyStoreFactory);
+  }
+
+  /**
+   * Constructor.
+   * @param rootPath The root path of the {@link HelixPropertyStore}.
+   * @param topicPath The path under which topic are created as {@link ZNRecord}s.
+   * @param zkClientConnectString The string to connect to a ZooKeeper server.
+   * @param zkClientSessionTimeoutMs The timeout in millisecond within which reconnection to the ZooKeeper server
+   *                                 will be considered as the same session.
+   * @param zkClientConnectionTimeoutMs The timeout in millisecond to establish a connection to the Zookeeper server.
+   * @param storeFactory The factory to start a {@link HelixPropertyStore}.
+   */
+  public HelixNotifier(String rootPath, String topicPath, String zkClientConnectString, int zkClientSessionTimeoutMs,
+      int zkClientConnectionTimeoutMs, HelixPropertyStoreFactory<ZNRecord> storeFactory) {
+    if (storeFactory == null) {
+      throw new IllegalArgumentException("storeFactory cannot be null.");
+    }
+    this.topicPath = topicPath;
+    List<String> subscribedPaths = new ArrayList<>();
+    subscribedPaths.add(rootPath + topicPath);
+    this.helixStore =
+        storeFactory.getHelixPropertyStore(zkClientConnectString, zkClientSessionTimeoutMs, zkClientConnectionTimeoutMs,
+            rootPath, subscribedPaths);
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * Return {@code true} does not guarantee all the {@link TopicListener} will receive the message. It just indicates
+   * the message has been successfully dent out.
+   */
+  @Override
+  public boolean sendMessage(String topic, String message) {
+    if (topic == null) {
+      throw new IllegalArgumentException("path to listen cannot be null");
+    }
+    if (message == null) {
+      throw new IllegalArgumentException("message cannot be null");
+    }
+    String topicPath = makeTopicPath(topic);
+    ZNRecord record = new ZNRecord(topicPath);
+    record.setSimpleField(MESSAGE_KEY, message);
+    return helixStore.set(topicPath, record, AccessOption.PERSISTENT);
```

Comment:
Concurrency will be handled by `HelixPropertyStore`, so if two messages arrive one after another, the last wins. `HelixPropertyStore` also has a retry mechanism, for example, if n updates come together, one will succeed, the rest will be retried, and eventually they will be written.